### DISCUSSION TITLE
Training from scratch

### DISCUSSION THREAD

**Discussion by zeleny-ondrej** [ultralytics/yolov5#7028] [Q&A]
Hello, 
im trying on sligtly modified version of Yolov5, however my project has different filenames and location of layers in those files, im trying to train the model to detect object without using pretrained weights (since it cant find it due to different location). However the issue is that im not even reaching 1% precision after 200 epochs of training. Did anyone managed to train the algorithm without using the pretrained weights ? Or is it possible to modify pretrained weights and change the location of layer classes (nn.Module).

**Comment by glenn-jocher** [comment]
@Zelda000 all official models are trained from scratch. Commands to reproduce are shown clearly in README, you should start there:

<img width="990" alt="Screenshot 2022-03-17 at 22 47 39" src="https://user-images.githubusercontent.com/26833433/158900255-41b70232-f51b-4822-838f-8286384972ea.png">

**Reply by zeleny-ondrej** [reply]
Thank you. I completle forgor about that. Did it really took 2 days of tranning to produce pretrained weights for yolov5s ? If im correct i see graphs of 300 epochs, what was the batchsize ? Since for my application im using only come classes of COCO, im not sure if i should run the training with it or make my own general weights  with COCO and then re-train to the particular application. What are your thoughts ?

**Reply by glenn-jocher** [reply]
@Zelda000 yes YOLOv5s trains 300 COCO epochs on a V100 in about 2 days. Batch size is indicated in above image.

Starting from pretrained weights is recommended for most use cases.

**Reply by zeleny-ondrej** [reply]
Thanks you. As it turn out, my dataset was not enough for the algoritm to learn fast enough. I tried it again with the whole COCO dataset, 64 batch size and im already getting reasonable precision after just few epochs. The reason i cant use pretrained weights is because i have defined the network differently, so the attempt_load() cant find  classes for layer modules (Conv, CSP,SPPF etc., located in your implementation modules/common.py).

**Reply by glenn-jocher** [reply]
Oh got it. Then you could train COCO yourself with your new architecture and use those as pretrained weights.

**Reply by zeleny-ondrej** [reply]
So, i've running the training algorithm for about 50 epochs on whole COCO dataset, but im having issues replicating your results. I have checked wandb for hyperparameters and i have the same values.
![image](https://user-images.githubusercontent.com/38864594/160234819-04dc15ba-d370-47a3-8717-ab47cfd4b73c.png)

When i tried to run training on my smaller dataset which is just subset of coco dataset with the use of weights from those 50 epochs, i ran intoi this issue.

  File "train.py", line 346, in <module>
    main(opt)
  File "train.py", line 335, in main
    Training(opt)
  File "train.py", line 218, in Training
    scaler.step(optimizer)  # optimizer.step
  File "/usr/local/lib/python3.7/dist-packages/torch/cuda/amp/grad_scaler.py", line 338, in step
    retval = self._maybe_opt_step(optimizer, optimizer_state, *args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch/cuda/amp/grad_scaler.py", line 285, in _maybe_opt_step
    retval = optimizer.step(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch/optim/lr_scheduler.py", line 65, in wrapper
    return wrapped(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch/optim/optimizer.py", line 88, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch/optim/sgd.py", line 143, in step
    nesterov=nesterov)
  File "/usr/local/lib/python3.7/dist-packages/torch/optim/_functional.py", line 173, in sgd
    buf.mul_(momentum).add_(d_p, alpha=1 - dampening)
RuntimeError: The size of tensor a (255) must match the size of tensor b (30) at non-singleton dimension 0

i dont see what the issue might be since the dataset is preaty much the same with just nc=5 instead of 80.
Do you have any tips how to resolve this ? The training works when i try to resume training on the whole COCO dataset.

**Reply by zeleny-ondrej** [reply]
Found the issue in optimizer.load_state_dict(), it seems it cant be used when re-training the model.

**Reply by glenn-jocher** [reply]
@Zelda000 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) to help us better understand and diagnose your problem. 

Thank you! 😃
